### PR TITLE
Compose FabLayout

### PR DESCRIPTION
**Pull request by VDostoyevskiy**
TODO
Suspend show && hide with mutator

### CODE REVIEW

File: app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/ComposeFabLayout.kt [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import com.hippo.ehviewer.ui.tools.animateFloatMergePredictiveBackAsState
+
+typealias OnExpandStateListener = (Boolean) -> Unit
+
+class ComposeFabLayout @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0,
+) : AbstractComposeView(context, attrs, defStyle) {
+    var isModeExpand by mutableStateOf(true)
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/ComposeFabLayout.kt:39]: I prefer something like `isExpandable`.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/ComposeFabLayout.kt [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    private val expandedBackingField = mutableStateOf(false)
+    private var expanded by expandedBackingField
+    private val listeners = mutableListOf<OnExpandStateListener>()
+
+    fun addOnExpandStateListener(listener: OnExpandStateListener) {
+        listeners.add(listener)
+    }
+
+    fun show() {
+        expanded = true
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/ComposeFabLayout.kt:54]: Should hide the FAB as well.

Comment (VDostoyevskiy) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/ComposeFabLayout.kt:54]: > Should not expand the FAB when showing.

When showing, expanded is true
So it's no-op

Comment (VDostoyevskiy) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/ComposeFabLayout.kt:54]: > Should not expand the FAB when showing.

Oops, show && hide is not implemented

File: app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/ComposeFabLayout.kt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    fun hide() {
+        expanded = false
+    }
+
+    @Composable
+    override fun Content() {
+        Mdc3Theme {
+            if (expanded) {
+                Spacer(
+                    modifier = Modifier.fillMaxSize().clickable(
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/ComposeFabLayout.kt:66]: This blocks scrolling.

**Comment by VDostoyevskiy** [comment]
@FooIbar PTAL

File: app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt [hunk 4]
```diff
@@ -1,482 +1,134 @@
...
-                            child.alpha = 1f
-                        }
-                    }
-                } else {
-                    if (mHidePrimaryFab) {
-                        setPrimaryFabAnimation(getChildAt(count - 1), expanded, !expanded)
+            val coroutineScope = rememberCoroutineScope()
+            val appearState by animateFloatAsState(
+                targetValue = if (hidden) 0f else 1f,
+                animationSpec = tween(300),
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt:90]: Better to not use magic number.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt [hunk 5]
```diff
@@ -1,482 +1,134 @@
...
+            )
+            Box(
+                modifier = Modifier.fillMaxSize().padding(16.dp),
+                contentAlignment = Alignment.BottomEnd,
+            ) {
+                Box(contentAlignment = Alignment.Center) {
+                    val animatedProgress by animateFloatMergePredictiveBackAsState(expandedBackingField)
+                    FloatingActionButton(
+                        onClick = { expanded = !expanded },
+                        modifier = Modifier.rotate(lerp(90f, 0f, appearState)).scale(appearState),
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt:101]: I believe it was counterclockwise for hiding before?

File: app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt (hunk 4 above)
Comment (VDostoyevskiy) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt (hunk 5 above)
Comment (VDostoyevskiy) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt]: Done

**Comment by FooIbar** [comment]
What's the thought behind changing the icons?

**Comment by VDostoyevskiy** [comment]
> What's the thought behind changing the icons and their orders?

Simply just that I cannot find correspond one in materialIcons and forget the order ...

**Comment by VDostoyevskiy** [comment]
> What's the thought behind changing the icons and their orders?

Now we can revert it back

**Comment by FooIbar** [comment]
> > What's the thought behind changing the icons and their orders?
> 
> Simply just that I cannot find correspond one in materialIcons and forget the order ...

There is the [last page](https://fonts.google.com/icons?icon.query=last+page) icon.
The closest for the go to is [u turn right](https://fonts.google.com/icons?icon.query=u+turn+right), maybe we need to create it.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt [hunk 6]
```diff
@@ -1,482 +1,138 @@
...
+                                        expanded = false
+                                    }
+                                },
+                                modifier = Modifier.layout { measurable, constraints ->
+                                    val placeable = measurable.measure(constraints)
+                                    layout(placeable.width, placeable.height) {
+                                        val distance = lerp(150 * (size - index) + 50, 0, animatedProgress)
+                                        placeable.placeRelative(0, -distance, -(size - index).toFloat())
+                                    }
+                                }.rotate(lerp(90f, 0f, appearState)).scale(appearState),
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt:126]: Should secondary FAB rotate & scale? It's kind of strange.

**Review by FooIbar — APPROVED** [review]

File: app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt (hunk 6 above)
Comment (VDostoyevskiy) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/legacy/FabLayout.kt:126]: I think it's cool

**Pull request merged**